### PR TITLE
Buffers n strings

### PR DESCRIPTION
**Pull request by jystic**
- Buffers are now always unboxed
- Constant strings are output using @amosr strings

### CODE REVIEW

File: data/sea/31-buffer.h [hunk 1]
```diff
@@ -112,101 +93,141 @@ Read(buf)
...
+                                                                                \
+    for (iint_t ix = 0; ix != min; ++ix) {                                      \
+        iint_t x_ix = (ix + x.head) % n;                                        \
+        iint_t y_ix = (ix + y.head) % n;                                        \
+                                                                                \
+        if (!t##_lt(x.vals[x_ix], y.vals[y_ix]))                                \
+            return ifalse;                                                      \
+    }                                                                           \
+                                                                                \
+    return x.size < y.size;                                                     \
```
Comment (amosr) [inline data/sea/31-buffer.h]: CONCAT(iint_, op) (x.size, y.size)
?

File: data/sea/00-includes.h [hunk 2]
```diff
@@ -60,3 +60,6 @@ STATIC_ASSERT(sizeof (void *) == sizeof (uint64_t), icicle_only_supports_systems
 #else
 #   define ICICLE_WHEN_DEBUG(code)
 #endif
+
+#define CONCAT0(x, y) x ## y
```
Comment (tranma) [inline data/sea/00-includes.h]: this fine without parens?

File: src/Icicle/Sea/FromAvalanche/Prim.hs [hunk 3]
```diff
@@ -20,48 +22,62 @@ import           Icicle.Sea.FromAvalanche.Type
...
 
+data ArgConv
+ = ArgVal
+ | ArgRef
+
 data PrimDoc
- = PDFun   Doc
- | PDAlloc Doc
+ = PDFun   Doc (Maybe [ArgConv])
+ | PDAlloc Doc (Maybe [ArgConv])
```
Comment (amosr) [inline src/Icicle/Sea/FromAvalanche/Prim.hs]: neat

File: data/sea/00-includes.h (hunk 2 above)
Comment (jystic) [inline data/sea/00-includes.h]: I needs to have no parens or they'll become part of the string

File: src/Icicle/Sea/FromAvalanche/Program.hs [hunk 4]
```diff
@@ -32,7 +32,7 @@ import           Icicle.Sea.FromAvalanche.Prim
 import           Icicle.Sea.FromAvalanche.State
 import           Icicle.Sea.FromAvalanche.Type
 
-import           P
+import           P hiding (head)
```
Comment (amosr) [inline src/Icicle/Sea/FromAvalanche/Program.hs]: hiding head?

Comment (jystic) [inline src/Icicle/Sea/FromAvalanche/Program.hs]: I wanted to use it as a variable name (for buffer)

Fwiw, I don't even know why we have head in P??

Comment (amosr) [inline src/Icicle/Sea/FromAvalanche/Program.hs]: I'm not sure. is it safe head? maybe it's just re-exported with a warning like error. hmm..

**Comment by amosr** [comment]
so beautiful :+1:

**Pull request merged**